cluster/rpc.go

```diff
@@ -3,50 +3,196 @@ package cluster
-	if err != nil {
-		return nil, fmt.Errorf("failed to dial service address %v: %w", dialAddr, err)
+func (c *Client) Call(serviceMethod string, args interface{}, reply interface{}) error {
+	client := c.conns.Get()
+	return client.Call(serviceMethod, args, reply)
+}
+
+func (c *Client) Go(serviceMethod string, args interface{}, reply interface{}, done chan *rpc.Call) *rpc.Call {
+	client := c.conns.Get()
+	return client.Go(serviceMethod, args, reply, done)
+}
+
+func (c *Client) Close() error {
+	return c.conns.Close()
+}
+
+func (c *Client) ConnectionErrs() chan error {
+	return c.conns.errChan
+}
+
+const defaultMaxConnections = 3
+const defaultInitalNodeTimeout = 5 * time.Second
+
+type connectionBalancer struct {
+	clientNode string
+	seq        uint64
+
+	selectedNodes []Node
+	clients       []*rpc.Client
+	lock          sync.RWMutex
+
+	cancel    func()
+	errChan   chan error
+	waitGroup sync.WaitGroup
+}
+
+func newConnectionBalancer(host string, serviceName string, r Registry) (*connectionBalancer, error) {
+	ctx, cancel := context.WithCancel(context.Background())
+
+	c := &connectionBalancer{
+		clientNode: host,
+		clients:    []*rpc.Client{},
+		cancel:     cancel,
+		errChan:    make(chan error, 1),
 	}
 
-	return &Client{
-		Client:   client,
-		registry: r,
-	}, nil
+	nodesChan := r.WatchService(ctx, serviceName)
+
+	var initalNodes []Node
+	select {
+	case initalNodes = <-nodesChan:
+	case <-time.After(defaultInitalNodeTimeout):
+		return nil, fmt.Errorf("no inital nodes provided for %v", serviceName)
+	}
+
+	if err := c.handleNewNodes(initalNodes); err != nil {
+		return nil, err
+	}
+
+	c.waitGroup.Add(1)
+	go c.watchForNewNodes(ctx, nodesChan)
+	return c, nil
 }
 
-func nodeToDial(serviceName string, r Registry) (node Node, err error) {
-	serv, err := r.Services(context.Background())
+func (c *connectionBalancer) Get() *rpc.Client {
+	return c.roundRobinSelect()
+}
+
+func (c *connectionBalancer) roundRobinSelect() *rpc.Client {
+	index := atomic.AddUint64(&c.seq, uint64(1))
+	return c.clients[int(index)%len(c.clients)]
+}
+
+func (c *connectionBalancer) Close() error {
+	if c.cancel != nil {
+		c.cancel()
+	}
+	for _, client := range c.clients {
+		client.Close()
+	}
+	c.waitGroup.Wait()
+	close(c.errChan)
+	return nil
+}
+
+func (c *connectionBalancer) watchForNewNodes(ctx context.Context, nodesChan chan []Node) {
+	defer c.waitGroup.Done()
+
+	for {
+		select {
+		case nodes := <-nodesChan:
+			if len(nodes) == 0 {
+				continue
+			}
+
+			if err := c.handleNewNodes(nodes); err != nil {
+				select {
+				case c.errChan <- err:
+				default:
+				}
+			}
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+func (c *connectionBalancer) handleNewNodes(nodes []Node) error {
+	selectedNodes := c.selectNodes(nodes, defaultMaxConnections)
+	clients, err := c.connectToNodes(selectedNodes)
 	if err != nil {
-		return node, err
+		return err
+	}
+
+	c.lock.Lock()
+	c.clients = clients
+	c.selectedNodes = selectedNodes
+	c.lock.Unlock()
```

Comment:
Anyone know a better way of handling the state changes without a lock? It's not a big deal, just curious if there's a more idomadic way